contracts/0.8.9/oracle/ValidatorsExitBusOracle.sol

```diff
@@ -0,0 +1,507 @@
+
+
+contract ValidatorsExitBusOracle is BaseOracle {
+    using RateLimit for RateLimit.State;
+    using UnstructuredStorage for bytes32;
+    using SafeCast for uint256;
+
+    error AdminCannotBeZero();
+    error SenderNotAllowed();
+    error UnsupportedRequestsDataFormat(uint256 format);
+    error InvalidRequestsData();
+    error InvalidRequestsDataLength();
+    error UnexpectedRequestsDataLength();
+    error InvalidRequestsDataSortOrder();
+    error ArgumentOutOfBounds();
+
+    event DataBoundariesSet(
+        uint256 indexed refSlot,
+        uint256 maxExitRequestsPerReport,
+        uint256 maxExitRequestsListLength,
+        uint256 exitRequestsRateLimitWindowSizeSlots,
+        uint256 exitRequestsRateLimitMaxThroughputE18
+    );
+
+    event ValidatorExitRequest(
+        uint256 indexed stakingModuleId,
+        uint256 indexed nodeOperatorId,
+        uint256 indexed validatorIndex,
+        bytes validatorPubkey,
+        uint256 timestamp
+    );
+
+    event WarnDataIncompleteProcessing(
+        uint256 indexed refSlot,
+        uint256 requestsProcessed,
+        uint256 requestsCount
+    );
+
+    struct DataBoundaries {
+        uint64 maxRequestsPerReport;
+        uint64 maxRequestsListLength;
+    }
+
+    struct DataProcessingState {
+        uint256 lastProcessedItemWithoutPubkey;
+        uint64 refSlot;
+        uint64 requestsCount;
+        uint64 requestsProcessed;
+        uint16 dataFormat;
+    }
+
+    struct RequestedValidator {
+        bool requested;
+        uint64 index;
+    }
+
+    /// @notice An ACL role granting the permission to submit the data for a committee report.
+    bytes32 public constant SUBMIT_DATA_ROLE = keccak256("SUBMIT_DATA_ROLE");
+
+    /// @notice An ACL role granting the permission to set report data safety boundaries.
+    bytes32 public constant MANAGE_DATA_BOUNDARIES_ROLE = keccak256("MANAGE_DATA_BOUNDARIES_ROLE");
+
+
+    /// @dev Storage slot: uint256 totalRequestsProcessed
+    bytes32 internal constant TOTAL_REQUESTS_PROCESSED_POSITION =
+        keccak256("lido.ValidatorsExitBusOracle.totalRequestsProcessed");
+
+    /// @dev Storage slot: mapping(uint256 => RequestedValidator) lastRequestedValidatorIndices
+    /// A mapping from the (moduleId, nodeOpId) packed key to the last requested validator index.
+    bytes32 internal constant LAST_REQUESTED_VALIDATOR_INDICES_POSITION =
+        keccak256("lido.ValidatorsExitBusOracle.lastRequestedValidatorIndices");
+
+    /// @dev Storage slot: RateLimit.State rateLimit
+    bytes32 internal constant RATE_LIMIT_POSITION =
+        keccak256("lido.ValidatorsExitBusOracle.rateLimit");
+
+    /// @dev Storage slot: DataBoundaries dataBoundaries
+    bytes32 internal constant DATA_BOUNDARIES_POSITION =
+        keccak256("lido.ValidatorsExitBusOracle.dataBoundaries");
+
+    /// @dev Storage slot: DataProcessingState dataProcessingState
+    bytes32 internal constant DATA_PROCESSING_STATE_POSITION =
```

Comment:
since the oracle apply data in one tx maybe let's remove this one?

Reply:
removing support for multi-tx data delivery will require re-writing significant part of the oracle code

Reply:
doubt we have time for that now